### ISSUE TITLE
IdentityHashMapUsage warnings differ from its documentation

### ISSUE THREAD

**Issue by mernst** [google/error-prone#2095]
Error Prone 2.5.0 issues `IdentityHashMapUsage` for code like this:

```java
  void f(Map<String, Integer> map) {
    new IdentityHashMap<String, Integer>(map);
  }
```

1. This pattern is not what is described at https://errorprone.info/bugpattern/IdentityHashMapUsage, which says
"This means that we shouldn’t type upcast IdentityHashMap to its super classes and always use the default type."
Please either:
    * relax the IdentityHashMapUsege rule to not forbid the above code, or
    * edit https://errorprone.info/bugpattern/IdentityHashMapUsage to describe the additional checks, and to explain why they are problematic.

2. I am having trouble parsing the last part of the sentence "and always use the default type".
    * Is that clause explaining a good or bad pattern?  Grammatically it is within the scope of "shouldn't" which suggests it is a pattern to avoid, but "always" suggests it is a pattern to use.  Please clarify the wording.
    * What does "the default type" mean in that clause?

3. In the upper right corner of the webpage, under "Severity WARNING" there is nothing after "Has Fix?".  I expected a boolean such as "Yes" or "No".  Please either add that, or remove the "Has Fix?" text.

**Comment by cushon** [comment]
Thanks! My recollection is that warning on `new IdentityHashMap<String, Integer>(map)` is deliberate, I agree that if we keep that behaviour the docs should cover it, and that the docs could use some work either way.

I think the general claim is that the semantics of `IdentityHashMap` and `Map` are different enough that they should be treated as different types, which includes both things like using the explicit `IdentityHashMap` type for variables and return types instead of just using `Map` where appropriate, and also avoiding converting existing `Map`s to `IdentityHashMap`s (as in the `new IdentityHashMap<String, Integer>(map)` example).

I don't think we had a lot of specific examples where stuff like `new IdentityHashMap<>(map)` lead to bugs, it just seems potentially suspicious and we also didn't have examples of it being done deliberately. If you have feedback based on the code where you encountered this check that'd also be welcome--did you have a use-case where something like `new IdentityHashMap<>(map)` was correct/deliberate?

(And I removed the obsolete "Has Fix?" text from the docs, thanks for noticing that.)

**Comment by mernst** [comment]
> I don't think we had a lot of specific examples where stuff like `new IdentityHashMap<>(map)` lead to bugs, it just seems potentially suspicious and we also didn't have examples of it being done deliberately.

Here are some examples:
 * https://github.com/typetools/checker-framework/blob/f3c9ecc242b879a3949a9e3163545be4eb23d86c/dataflow/src/main/java/org/checkerframework/dataflow/analysis/AnalysisResult.java#L87
 * https://github.com/typetools/checker-framework/blob/afaa95f84d1c7116ed0a226bd3b3dc24d1b3c760/framework/src/main/java/org/checkerframework/framework/type/AnnotatedTypeCopierWithReplacement.java#L40

I don't see how this pattern is harmful.  If the pattern is not harmful (does not lead to known bugs), I think it would be better if Error Prone does not issue warnings about it (yet, until there is evidence that it is harmful).

**Comment by cushon** [comment]
I can't find any usages of the second example, but in the first example it looks like the arguments to those parameters are `IdentityHashMap`, i.e. it would be possible to change the formal parameter types to `IdentityHashMap`?

https://github.com/typetools/checker-framework/blob/f34b1b8812b4e58d73cfb6e3d81938ab92e67e55/dataflow/src/main/java/org/checkerframework/dataflow/analysis/AbstractAnalysis.java#L173-L175

The idea is that building up e.g. a `LinkedHashMap` and then passing it to that constructor of `IdentityHashMap` is usually going to be a mistake. I think it's subjective whether that's likely enough to cause problems in practice for it to be worth strongly preferring to type things as `IdentityHashMap` instead of `Map` to prevent that mistake, but it seems like the check is basically doing what it was intended to do here.

**Comment by mernst** [comment]
OK, I can propagate the `IdentityHashMap` backward to other uses.  Thanks for looking into this.

It would be great to update the documentation to reflect the checks that are being performed.
